### PR TITLE
kaustavha/Refactor/hostinfo

### PR DESCRIPTION
**Pull request by kaustavha**
Refactor hostinfo checks to increase readability, reduce footprint, increase testability and reliability.

### CODE REVIEW

File: hostinfo/autoupdates.lua [hunk 1]
```diff
@@ -15,71 +15,85 @@ limitations under the License.
...
+        status = status
+      })
+    end
+    self:_pushParams(errTable, outTable)
     return callback()
   end
-  return execFileToBuffers(statcmd, statargs, options, statExecCb)
+
+  -- Export internals for testing
+  exports.options = options
```
Comment (rphillips) [inline hostinfo/autoupdates.lua]: exports should not be modified within the class instance.

File: hostinfo/autoupdates.lua [hunk 2]
```diff
@@ -15,71 +15,85 @@ limitations under the License.
...
+  exports.onLine = onLine
+  exports.onErr = onErr
+  exports.finalCb = finalCb
+
+  return streamingSpawn(cmd, args, opts, onLine, onErr, finalCb)
```
Comment (rphillips) [inline hostinfo/autoupdates.lua]: return is superfluous here

File: hostinfo/autoupdates.lua [hunk 3]
```diff
@@ -15,71 +15,85 @@ limitations under the License.
...
-    method = 'yum_cron'
-    options = {}
-  else
-    self._error = 'Could not determine linux distro for autoupdates'
+  local spawnConfig = getInfoByVendor(options)
+  if not spawnConfig.cmd then
+    self._error = string.format("Couldn't decipher linux distro for check %s",  self:getType())
     return callback()
   end
+  cmd, args, opts = spawnConfig.cmd, spawnConfig.args, {}
```
Comment (rphillips) [inline hostinfo/autoupdates.lua]: makes more sense to put the local on the same line as these args

File: hostinfo/base.lua [hunk 4]
```diff
@@ -14,12 +14,15 @@ See the License for the specific language governing permissions and
...
+local gmtNow = require('virgo/utils').gmtNow
+local tableToString = require('virgo/util/misc').tableToString
+local los = require('los')
+-------------------------------------------------------------------------------
 
---[[ HostInfo ]]--
-local HostInfo = Object:extend()
+local HostInfo = Transform:extend()
 function HostInfo:initialize()
+  Transform.initialize(self, {objectMode = true})
```
Comment (rphillips) [inline hostinfo/base.lua]: hostinfo does not need to be a transform if HostInfo is not a transform

File: hostinfo/base.lua [hunk 5]
```diff
@@ -28,11 +31,53 @@ function HostInfo:serialize()
...
+    return true
+  end
+  for _, platform in pairs(self:getPlatforms()) do
+    if platform == currentPlatform then
+      return true
+    end
+  end
+end
+
+function HostInfo:_pushParams(err, data)
```
Comment (rphillips) [inline hostinfo/base.lua]: _pushError and _pushParams

File: hostinfo/cron.lua [hunk 6]
```diff
@@ -15,48 +15,48 @@ limitations under the License.
 --]]
 local HostInfo = require('./base').HostInfo
 
-local los = require('los')
-local readCast = require('./misc').readCast
+local m = require('./misc')
```
Comment (rphillips) [inline hostinfo/cron.lua]: should be misc.

File: hostinfo/cron.lua [hunk 7]
```diff
@@ -15,48 +15,48 @@ limitations under the License.
...
-  local cdir, vendor, errTable
-  vendor = sigar:new():sysinfo().vendor:lower()
-  errTable = {}
-  if vendor == 'ubuntu' or vendor == 'debian' then
-    cdir = '/var/spool/cron/crontabs'
-  elseif vendor == 'rhel' or vendor == 'centos' then
-    cdir = '/var/spool/cron'
-  else
-    self._error = 'Could not determine OS'
+  deb = {dir = '/var/spool/cron/crontabs' }
```
Comment (rphillips) [inline hostinfo/cron.lua]: local deb and local rhel and local options

try and keep the locals close to together.

File: hostinfo/cron.lua [hunk 8]
```diff
@@ -67,23 +67,34 @@ function Info:run(callback)
...
+    -- We wanna be able to return empty sets for empty crontabs therefore we dont use self:_pushparams
+    if next(errTable) and not next(outTable) then
+      self._error = table.concat(errTable, ' ')
+    end
+    self._params = outTable
+    return callback()
+  end
+
+  -- surface internals for testing
+  exports.options = options
```
Comment (rphillips) [inline hostinfo/cron.lua]: exports should not be set here.

File: hostinfo/cve.lua [hunk 9]
```diff
@@ -15,62 +15,71 @@ limitations under the License.
 --]]
 local HostInfo = require('./base').HostInfo
 
-local table = require('table')
-local execFileToBuffers = require('./misc').execFileToBuffers
-local los = require('los')
-local sigar = require('sigar')
-local fmt = require('string').format
+local m = require('./misc')
```
Comment (rphillips) [inline hostinfo/cve.lua]: should be misc

File: hostinfo/misc.lua [hunk 10]
```diff
@@ -55,56 +59,98 @@ local function execFileToBuffers(command, args, options, callback)
...
-    else
-      table.insert(errTable, 'file not found')
-      return callback()
+local function wrappedExecFileToStreams(command, arguments, options, onLine, onErr, callback)
+  -- Make onErr an optional parameter
+  if not callback and onErr then
+    callback = onErr
+    onErr = function() return end
+  end
+  assert(command, 'Parameter missing: (command) Command to spawn childprocess with is required')
```
Comment (rphillips) [inline hostinfo/misc.lua]: we should take the asserts out

File: hostinfo/misc.lua [hunk 11]
```diff
@@ -115,22 +161,64 @@ local function asyncSpawn(dataArr, spawnFunc, successFunc, finalCb)
...
+    end
+    local function onClose(_exitCode)
+      exitCode = _exitCode
+      done()
     end
-    local cmd, args = spawnFunc(datum)
-    return execFileToBuffers(cmd, args, opts, _successFunc)
+
+    cmd, args = spawnFunc(datum)
+    execFileToBuffers('which', {cmd}, {env = process.env}, function(err, ec, stdout, stderr)
```
Comment (rphillips) [inline hostinfo/misc.lua]: should not need `which`

File: hostinfo/misc.lua [hunk 12]
```diff
@@ -115,22 +161,64 @@ local function asyncSpawn(dataArr, spawnFunc, successFunc, finalCb)
...
+      if stdout then cmd = trim(stdout) end
+      child, stdout, stderr = execFileToStreams(cmd,
+        args,
+        { env = process.env })
+      child:once('close', onClose)
+      stdout
```
Comment (rphillips) [inline hostinfo/misc.lua]: ```stdout:on('data'...```

File: hostinfo/misc.lua [hunk 13]
```diff
@@ -115,22 +161,64 @@ local function asyncSpawn(dataArr, spawnFunc, successFunc, finalCb)
...
+  if options.default then return options.default end
+  local NilInfo = require('./nil')
+  return NilInfo
+end
+
+
+exports.execFileToBuffers = execFileToBuffers
+exports.execFileToStreams = execFileToStreams
+exports.wrappedExecFileToStreams = wrappedExecFileToStreams
+exports.streamingSpawn = wrappedExecFileToStreams
```
Comment (rphillips) [inline hostinfo/misc.lua]: ```exports.streamingSpawn = streamingSpawn``` or needs to be removed

**Comment by rphillips** [comment]
Failed here: https://travis-ci.org/virgo-agent-toolkit/rackspace-monitoring-agent/builds/76027321#L1236

File: hostinfo/autoupdates.lua [hunk 14]
```diff
@@ -14,71 +14,100 @@ See the License for the specific language governing permissions and
...
+      update_method = 'unattended_upgrades',
+      status = 'enabled'
+    })
+  end
+  cb()
+end
+
+local YumReader = Reader:extend()
+function YumReader:_transform(line, cb)
+  if line then self:push({
```
Comment (rphillips) [inline hostinfo/autoupdates.lua]: line should always be non nil

File: hostinfo/base.lua [hunk 15]
```diff
@@ -47,4 +54,55 @@ function HostInfo:run(callback)
...
+  if not self:getPlatforms() then
+    return true
+  elseif #self:getPlatforms() == 0 then
+    return true
+  end
+  for _, platform in pairs(self:getPlatforms()) do
+    if platform == currentPlatform then
+      return true
+    end
+  end
```
Comment (rphillips) [inline hostinfo/base.lua]: should return false here to make it return a boolean

File: hostinfo/cpu.lua [hunk 16]
```diff
@@ -57,6 +58,7 @@ function Info:initialize()
     obj['name'] = name
     table.insert(self._params, obj)
   end
+  return callback()
```
Comment (rphillips) [inline hostinfo/cpu.lua]: return isn't needed here

File: hostinfo/cron.lua [hunk 17]
```diff
@@ -15,81 +15,114 @@ limitations under the License.
...
+  self:push({
+    time = time,
+    m = m,
+    h = h,
+    dom = dom,
+    mon = mon,
+    dow = dow,
+    command = command
+  })
+  return cb()
```
Comment (rphillips) [inline hostinfo/cron.lua]: return on this line isn't needed.

File: hostinfo/cron.lua [hunk 18]
```diff
@@ -15,81 +15,114 @@ limitations under the License.
...
 end
 
 function Info:_run(callback)
-  if los.type() ~= 'linux' then
-    self._error = 'Unsupported OS for pluggable auth module definitions'
-    return callback()
-  end
+  local errTable, outTable = {}, {}
+  local deb, rhel
+  deb = {dir = '/var/spool/cron/crontabs' }
```
Comment (rphillips) [inline hostinfo/cron.lua]: put local dev, local rhel on the same line

File: hostinfo/cve.lua [hunk 19]
```diff
@@ -15,62 +15,84 @@ limitations under the License.
...
 end
 
 function Info:_run(callback)
-  if los.type() ~= 'linux' then
-    self._error = 'Unsupported OS for pluggable auth module definitions'
-    return callback()
-  end
+  local errTable, outTable = {}, {}
+  local deb, rhel
+  deb = {cmd = '/bin/sh', args = {'-c',  'zcat /usr/share/doc/*/changelog.Debian.gz | grep CVE-'} }
```
Comment (rphillips) [inline hostinfo/cve.lua]: local deb = ...
local rhel = ...

File: hostinfo/init.lua [hunk 20]
```diff
@@ -60,14 +66,18 @@ end
...
-    local info = create(v)
-    local obj = info:serialize().metrics
-    data = data .. '-- ' .. v .. '.' .. os.type() .. ' --\n\n'
-    data = data .. misc.toString(obj)
-    data = data .. '\n'
-  end
-  fs.writeFile(fileName, data, callback)
+  async.forEachLimit(HOST_INFO_TYPES, 5, function(v, cb)
+    local klass = create(v)
+    klass:run(function(err, callback)
```
Comment (rphillips) [inline hostinfo/init.lua]: callback() here isn't used... I don't think it's needed.

File: hostinfo/misc.lua [hunk 21]
```diff
@@ -14,109 +14,86 @@ See the License for the specific language governing permissions and
...
-  end, function()
-    return finalCb(obj)
-  end)
-end
 
+local function getInfoByVendor(options)
+  local sysinfo = sigar:new():sysinfo()
+  local vendor = sysinfo.vendor:lower()
+  local name = sysinfo.name:lower()
+  if vendor == 'red hat' then vendor = 'rhel' end
```
Comment (rphillips) [inline hostinfo/misc.lua]: hmm. Is it two words 'red hat' ?

File: hostinfo/packages.lua [hunk 22]
```diff
@@ -14,64 +14,82 @@ See the License for the specific language governing permissions and
...
 end
 
 function Info:_run(callback)
-  if los.type() ~= 'linux' then
-    self._error = 'Unsupported OS for Packages'
-    return callback()
-  end
+  local errTable, outTable = {}, {}
+  local deb, rhel
+  deb = {cmd = 'dpkg-query', args = {'-W'}}
```
Comment (rphillips) [inline hostinfo/packages.lua]: local deb... Let's keep the format:

deb = {
  cmd = 'dpkg-query',
  args = {'-W'}
}

**Comment by rphillips** [comment]
It would be awesome to write a few tests as well.

File: hostinfo/misc.lua (hunk 21 above)
Comment (kaustavha) [inline hostinfo/misc.lua]: yes.  vendor returns "Red Hat" 
I added
```
  local sysinfo = sigar:new():sysinfo()
  local vendor = sysinfo.vendor:lower()
  local name = sysinfo.name:lower()
 print(name)
 print(vendor)
```
to a random hostinfo check and ran it, heres a screenie. Latest version of red hat we offer. 
screen shot 2015-08-20 at 5 01 09 pm

File: hostinfo/base.lua (hunk 15 above)
Comment (kaustavha) [inline hostinfo/base.lua]: Done

File: hostinfo/autoupdates.lua [hunk 23]
```diff
@@ -14,71 +14,100 @@ See the License for the specific language governing permissions and
...
-local table = require('table')
-local execFileToBuffers = require('./misc').execFileToBuffers
+local Reader = Transform:extend()
+function Reader:initialize()
+  Transform.initialize(self, {objectMode = true})
+  self._pushed = false
+end
+
+local AptReader = Reader:extend()
+function AptReader:_transform(line, cb)
```
Comment (rphillips) [inline hostinfo/autoupdates.lua]:
```lua
if self._pushed then return callback() end
```

File: tests/test-hostinfo.lua [hunk 24]
```diff
@@ -36,4 +42,126 @@ require('tap')(function(test)
...
+    reader:on('data', function(data) table.insert(outTable, data) end)
+    reader:once('end', function()
+      local outFixtureTable = json.parse(outFixture)
+      assert(is_table_equal(outFixtureTable, outTable), 'not ok: outFixture and outTable dont match')
+      assert(#errTable==0, 'Not ok: errtable is not 0 length')
+    end)
+    local chunk = hostinfoFixtureDir['packages_linux_in.txt']
+    inFixture:write(chunk)
+    inFixture:write(nil)
+  end)
```
Comment (rphillips) [inline tests/test-hostinfo.lua]: nice job... as I explained in IRC, let's tweak these test functions to use 'expect' on some of the callbacks, namely the ones for 'end'.

**Comment by rphillips** [comment]
+1

**Pull request merged**